tests/test_config.py

```diff
@@ -0,0 +1,314 @@
+    """Check errors in finding module schema."""
+
+    # Check that find_schema fails as expected
+    with pytest.raises(expected_exception):
+        config.find_schema(config_dict)
+
+    log_check(caplog, expected_log_entries)
+
+
+@pytest.mark.parametrize(
+    "modules,expected_exception,expected_log_entries",
+    [
+        (
+            ["a_stupid_module_name"],
+            RuntimeError,
+            (
+                (
+                    CRITICAL,
+                    "Expected a schema for a_stupid_module_name module configuration, "
+                    "it was not provided!",
+                ),
+            ),
+        ),
+    ],
+)
+def test_construct_combined_schema(
+    caplog, modules, expected_exception, expected_log_entries
+):
+    """Checks errors for bad or missing json schema."""
+
+    # Check that construct_combined_schema fails as expected
+    with pytest.raises(expected_exception):
+        config.construct_combined_schema(modules)
+
+    log_check(caplog, expected_log_entries)
+
+
+@pytest.mark.parametrize(
+    "expected_log_entries",
+    [
+        (
+            (
+                (
+                    INFO,
+                    "Configuration files successfully validated!",
+                ),
+                (
+                    INFO,
+                    "Saving all configuration details to ./complete_config.toml",
+                ),
+            )
+        ),
+    ],
+)
+def test_final_validation_log(caplog, expected_log_entries):
+    """Checks that validation passes as expected and produces the correct output."""
+
+    config.validate_config(["tests/fixtures"], out_file_name="complete_config")
+
+    # Remove generated output file
+    # As a bonus tests that output file was generated correctly + to the right location
+    Path("./complete_config.toml").unlink()
+
+    print(caplog)
+    # Then check that the correct (critical error) log messages are emitted
+    log_check(caplog, expected_log_entries)
+
+    # Check that final config has been within nested dictionary
+    assert type(config.COMPLETE_CONFIG["config"]) == dict
+
+
+# NOT SURE HOW TO CATCH LOGGED OUTPUT IN THIS CASE
+# TODO - EXTEND THIS TO ALSO CHECK LOGGER OUTPUT
+@pytest.mark.parametrize(
+    "schema_name,schema, expected_exception",
+    [
+        ("core", {}, ValueError),
+        (
+            "test",
+            "najsnjasnda",
+            OSError,
+        ),
+        (
+            "bad_module_1",
+            {"type": "object", "propertie": {"bad_module_1": {}}},
+            KeyError,
+        ),
+        (
+            "bad_module_2",
+            {"type": "object", "properties": {"bad_module_1": {}}},
+            KeyError,
+        ),
+        (
+            "bad_module_3",
+            {"type": "object", "properties": {"bad_module_3": {}}},
+            KeyError,
+        ),
+    ],
+)
+def test_register_schema_errors(schema_name, schema, expected_exception):
```

Comment:
@dalonsoa  I also had a quick question on this test. I could work out how to test that the decorator raised the correct error type in response to bad input, but couldn't work out how to check the error message in the log (which would contain info on what key was missing). I was wondering if you had any advice/pointers here?

Reply:
You want to capture the exception:

```
with pytest.raises(expected_exception) as err:
```

And then you can use `str(err)`.

Reply:
Hmmm `str(err)` doesn't seem to return a message here instead it returns `'<ExceptionInfo for raises contextmanager>'`, I think something about testing a decorator throws it off. `caplog` also didn't behave as I expected it to here

Reply:
Comparing `caplog` works only if it's not contained by the `with:` statement, but that is a workable solution